arch/x86/kvm/pkvm/pgtable.h

```diff
@@ -0,0 +1,237 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#ifndef __PKVM_X86_PGTABLE_H
+#define __PKVM_X86_PGTABLE_H
+
+#include <linux/types.h>
+#include <asm/pgtable_types.h>
+
+struct pkvm_pgtable;
+
+/*
+ * struct pkvm_pgtable_mm_ops	Page table memory management callbacks.
+ * @phys_to_virt:		Translate a virtual address to its physical
+ *				address.
+ * @virt_to_phys:		Translate a physical address to its virtual
+ *				address.
+ * @zalloc_page:		Allocate a page with clearing its contents.
+ * @get_page:			Increment the reference count of a given page.
+ * @put_page:			Decrement the reference count of a given page.
+ * @page_count:			Get the reference count of a given page.
+ * @flush_cache:		Invalidate the CPU cache for a given virtual
+ *				memory range.
+ */
+struct pkvm_pgtable_mm_ops {
+	void *(*phys_to_virt)(phys_addr_t phys);
+	phys_addr_t (*virt_to_phys)(void *vaddr);
+	void *(*zalloc_page)(void);
+	void (*get_page)(void *vaddr);
+	void (*put_page)(void *vaddr);
+	int (*page_count)(void *vaddr);
+	void (*flush_cache)(void *vaddr, unsigned int size);
+};
+
+/*
+ * struct pkvm_pgtable_ops	Page table operation callbacks.
+ * @pte_present			Detect a present pte.
+ * @pte_huge			Detect a huge pte.
+ * @pte_mkhuge			Set huge for the given pte.
+ * @pte_to_phys			Decode the physical address from pte.
+ * @pte_to_prot			Decode the property bits from pte.
+ * @calc_pte_prot		Calculate the property bits according to perm
+ *				arguments and page cache mode.
+ * @vaddr_to_index		Calculate the entry index for a virtual address
+ *				according to the given page table level.
+ * @level_to_size		Calculate the page size at a page table level.
+ * @level_to_mask		Calculate the page size mask at a page table
+ *				level.
+ * @pte_is_leaf			Detect a leaf pte.
+ * @pte_size			Calculate the pte size in bytes at a given page
+ *				table level.
+ * @pte_count			Calculate the number of ptes at a given page
+ *				table level.
+ * @set_pte			Set a pte to a given raw value.
+ * @get_pte			Get the raw value stored in the pte.
+ * @flush_tlb			Flush the TLB for a virtual address range.
+ */
+struct pkvm_pgtable_ops {
+	bool (*pte_present)(void *ptep);
+	bool (*pte_huge)(void *ptep);
+	void (*pte_mkhuge)(void *ptep);
+	unsigned long (*pte_to_phys)(void *ptep);
+	u64 (*pte_to_prot)(void *ptep);
+	u64 (*calc_pte_prot)(bool read, bool write, bool exec,
+			     enum page_cache_mode pc);
+	int (*vaddr_to_index)(unsigned long vaddr, int level);
+	unsigned long (*level_to_size)(int level);
+	u64 (*level_to_mask)(int level);
+	bool (*pte_is_leaf)(void *ptep, int level);
+	int (*pte_size)(int level);
+	int (*pte_count)(int level);
+	void (*set_pte)(void *ptep, u64 val);
```

Comment:
The `set_pte` will be conflict with the macro `set_pte` defined in asm/pgtable.h

With undef CONFIG_PARAVIRT_XXL, asm/pgtable.h will have a macro `#define set_pte(ptep, pte)              native_set_pte(ptep, pte)` defined. Then will have below compiling error with clang:

../arch/x86/kvm/pkvm/pgtable.c:52:15: error: no member named 'native_set_pte' in 'struct pkvm_pgtable_ops'
   52 |         pgt->pgt_ops.set_pte(ptep, pte);
      |         ~~~~~~~~~~~~ ^
../arch/x86/include/asm/pgtable.h:68:29: note: expanded from macro 'set_pte'
   68 | #define set_pte(ptep, pte)              native_set_pte(ptep, pte)

To fix this, we can use `(pgt->pgt_ops.set_pte)(ptep, pte)`, or probably better to give a different name e.g., write_pte/read_pte?

Reply:
SGTM.